caikit/runtime/servicers/global_predict_servicer.py

```diff
@@ -229,6 +234,179 @@ def Predict(
+                A Caikit Library data model response object
+        """
+        # Make sure the request has a model before doing anything
+        model_id = get_metadata(context, MODEL_MESH_MODEL_ID_KEY)
+        request_name = caikit_rpc.request.name
+
+        with alog.ContextLog(
+            log.debug,
+            "GlobalBackgroundPredictServicer.BackgroundPredict:%s",
+            request_name,
+        ):
+            # Before retrieving the model, which can trigger lazy backend
+            # initialization, we notify all backends of the context for this
+            # request which may update how the discovery logic works.
+            self.notify_backends_with_context(model_id, context)
+
+            # Retrieve the model from the model manager
+            log.debug("<RUN52259029D>", "Retrieving model '%s'", model_id)
+            model = self._model_manager.retrieve_model(model_id)
+            model_class = type(model)
+
+            # Little hackity hack: Calling _verify_model_task upfront here as well to
+            # short-circuit requests where the model is _totally_ unsupported
+            self._verify_model_task(model)
+
+            # Unmarshall the request object into the required module run argument(s)
+            with PREDICT_FROM_PROTO_SUMMARY.labels(
+                grpc_request=request_name, model_id=model_id
+            ).time():
+                inference_signature = model_class.get_inference_signature(
+                    input_streaming=caikit_rpc.input_streaming,
+                    output_streaming=caikit_rpc.output_streaming,
+                    task=caikit_rpc.task,
+                )
+                if not inference_signature:
+                    raise CaikitRuntimeException(
+                        StatusCode.INVALID_ARGUMENT,
+                        f"Model class {model_class} does not support {caikit_rpc.name}",
+                    )
+                if caikit_rpc.input_streaming:
+                    caikit_library_request = self._build_caikit_library_request_stream(
+                        request, inference_signature, caikit_rpc
+                    )
+                else:
+                    caikit_library_request = build_caikit_library_request_dict(
+                        request,
+                        inference_signature,
+                    )
+
+            response = self.run_prediction_job(
+                request_name,
+                model_id,
+                task=caikit_rpc.task,
+                prediction_func_name=inference_signature.method_name,
+                context=context,
+                context_arg=inference_signature.context_arg,
+                model=model,
+                wait=False,
+                **caikit_library_request,
+            )
+
+            # Don't measure
+            response_proto = build_proto_response(response)
+            return response_proto
+
+    def run_prediction_job(
```

Comment:
This function looks like it has a lot of duplicate code with `run_training_job`. Is there a refactor that could slim this down?

Reply:
Not really since most of the boiler plate is in the tracing/future creation. They both have their own validation and content of args/kwargs